lib/datasource.js

```diff
@@ -538,24 +525,112 @@ DataSource.prototype.defineRelations = function(modelClass, relations) {
       if ((targetModel && !isModelDataSourceAttached(targetModel)) ||
           (throughModel && !isModelDataSourceAttached(throughModel))) {
         // Create a listener to defer the relation set up
-        createListener(rn, r, targetModel, throughModel);
+        deferRelationSetup(relationName, r, targetModel, throughModel);
       } else {
         // The target model is resolved
         var params = traverse(r).clone();
-        params.as = rn;
-        params.model = polymorphicName || targetModel;
+        params.as = relationName;
+        params.model = targetModel;
         if (throughModel) {
           params.through = throughModel;
         }
-        modelClass[r.type].call(modelClass, rn, params);
+        modelClass[r.type].call(modelClass, relationName, params);
       }
     });
   }
 };
 
+function validateRelation(relationName, relation) {
+  var rn = relationName;
+  var r = relation;
+  var code;
+
+  // allows for method injection via relation definition in unit tests
+  getRelationValidationMsg = r.getRelationValidationMsg || getRelationValidationMsg;
+
```

Comment:
@jannyHou I reviewed your proposal. My point is actually precisely to avoid testing the exact message values, as they can further change for editorial reasons. I believe it is much less fragile to test the error **code** and the impacted relation (name and type) than to test the full error message, hence the mockup function. We did this once with Miroslav on Loopback.

A intermediate solution is to watermark the error code and relation type/name in the error object right in the `validateRelation()` method and assert them in the unit tests, but as it does not bring much functional added value to Datasource-juggler as these errors will pop at boot time and not at run time, I decided not to go this way.

Maybe you could summarize your cons against injecting a mocked method in the unit tests?
Would you prefer we watermark the params in the error object?